Added pagination.

JakeWengroff opened the pull request:
Hi @chalmagean,

I added pagination via the **will_paginate** rubygem. Not sure if this is the route you wanted to go, but I guess it does the job.

Jake

JakeWengroff committed:
Added pagination.

chalmagean commented:
That's awesome, but let's use the [kaminari gem](https://github.com/amatsuda/kaminari) instead of will paginate because it's actively maintained.

Also please add a some tests (in `spec/features`) around that enhancement.

JakeWengroff committed:
Changed pagination gem to kaminari.

JakeWengroff committed:
Added RSpec tests for Kaminari.

JakeWengroff commented:
Hi, Cezar,

I changed the pagination gem to Kaminari, and wrote a Controller test in RSpec, but it's still not passing.
I think it has to do with the Controller, not the test.
Let's discuss.

Jake

chalmagean commented on Gemfile:
```diff
@@ -10,6 +10,9 @@ gemspec
 # your gem to rubygems.org.
 
+gem 'kaminari'
+gem 'kaminari-rspec'
```
this one should probably go in the test group

JakeWengroff replied:
Agreed, I added **kaminari-rspec** to the test group.

chalmagean commented on Rakefile:
```diff
@@ -32,3 +32,6 @@ end
 
 task default: :test
+
+task 'test:prepare'
```
any reason for these two?

JakeWengroff replied:
I added these because I was getting an error, but I can remove them.

chalmagean commented on app/controllers/blog_posts_controller.rb:
```diff
@@ -2,9 +2,11 @@ class BlogPostsController < ApplicationController
 
   before_filter :find_tag!, :only => :filter
   before_filter :find_blog_post!, :only => :show
+  # default_scope order('published_at')
```
we should keep the default scope

JakeWengroff replied:
So this can be deleted?

chalmagean commented on app/controllers/blog_posts_controller.rb:
```diff
@@ -2,9 +2,11 @@ class BlogPostsController < ApplicationController
   def index
-    @blog_posts = BlogPost.all
+    @blog_posts = BlogPost.page(params[:page]).per(5)
+    # @blog_posts = BlogPost.order('published_at').page(params[:page]).per(5)
```
please don't commit commented code

JakeWengroff replied:
Got it.

chalmagean commented on app/views/blog_posts/index.html.erb:
```diff
@@ -14,6 +14,8 @@
         <% end %>
       </div>
     <% end %>
+  <% paginate @blog_posts %>
```
indentation seems off here... and there's an extra blank line

JakeWengroff replied:
Got it, fixed.

chalmagean commented on config/routes.rb:
```diff
@@ -6,7 +6,10 @@
   namespace :admin do
     resources :blog_posts do
       get "/get_tags" => "blog_posts#get_tags", as: :get_tags, on: :collection
+      get 'page/:page', :action => :index, :on => :collection
```
I don't think we want the `page` part in the url. Something like `/blog/2` should be fine

JakeWengroff replied:
I added this from the Kaminari documentation, https://github.com/amatsuda/kaminari#creating-friendly-urls-and-caching.

chalmagean replied:
Sure, it's just there's no point of having the extra word "page" in there.

JakeWengroff replied:
I understand - got it.

JakeWengroff replied:
Fixed.

chalmagean commented on spec/controllers/blog_posts_controller_spec.rb:
```diff
@@ -22,6 +22,12 @@
       expect(response).to render_template('blog_posts/index')
     end
+
+    it "paginates" do
```
Let's not do controller tests at all... let's make this a feature test instead.

JakeWengroff replied:
Ah, I didn't realize. (I have never created a feature test, so this will be my first one.)

I made this a feature test inside the `spec/features` folder, `paginate_posts_index_spec.rb`, which seems to be passing, but Kaminari's `paginate` method is now throwing an error across other tests:

```
NoMethodError:
     #   undefined method `total_pages' for #<BlogPost::ActiveRecord_Relation:0x007ff16a2a8038>
```

I can work on this.

chalmagean commented on spec/dummy/db/schema.rb:
```diff
@@ -13,12 +13,12 @@
 ActiveRecord::Schema.define(version: 20150115103932) do
 
-  create_table "blog_images", force: true do |t|
+  create_table "blog_images", force: :cascade do |t|
```
not sure why this is changed...

JakeWengroff replied:
I'm not sure either, I didn't add another controller.

chalmagean replied:
This is why: http://edgeguides.rubyonrails.org/4_2_release_notes.html#active-record-notable-changes

```
SchemaDumper uses force: :cascade on create_table. This makes it possible to reload a schema when foreign keys are in place.
```

chalmagean commented on spec/features/list_posts_spec.rb:
```diff
@@ -55,4 +55,5 @@
       visit filter_posts_path(:tag => "not_a_tag")
     }.to raise_exception(ActiveRecord::RecordNotFound)
   end
+
```
extra line

JakeWengroff replied:
A bad habit of mine. ;)

chalmagean commented on spec/spec_helper.rb:
```diff
@@ -14,6 +14,7 @@
 # Checks for pending migrations before tests are run.
 # If you are not using ActiveRecord, you can remove this line.
 ActiveRecord::Migration.check_pending! if defined?(ActiveRecord::Migration)
+# ActiveRecord::Migration.maintain_test_schema!
```
please don't commit commented code

JakeWengroff replied:
Got it.

chalmagean commented on spec/spec_helper.rb:
```diff
@@ -44,4 +45,5 @@
   config.order = "random"
   config.include FactoryGirl::Syntax::Methods
   config.include Rails.application.routes.url_helpers
+  # config.include KaminariRspec::TestHelpers, :type => :controller
```
please don't commit commented code

JakeWengroff replied:
Got it.

JakeWengroff committed:
Added separate Feature test.

I added a paginate_posts_index_spec.rb feature test to the spec/features folder, which seems to be passing, but Kaminari's paginate method seems to be throwing errors across other tests.

JakeWengroff committed:
Made changes to the Views, to make the tests pass.

I made changes to the Views, and ALL of the Features tests are passing (!!!)

chalmagean commented on app/controllers/blog_posts_controller.rb:
```diff
@@ -4,7 +4,7 @@ class BlogPostsController < ApplicationController
 
   def index
-    @blog_posts = BlogPost.all
+    @blog_posts = BlogPost.all.page(params[:page]).per(5)
```
Do we really need to call `.all` here?

chalmagean replied:
We should probably have more than 5 posts on a page (`.per(5)`). Also you should put the number in a constant.

JakeWengroff replied:
Fixed. No need to call `.all`, and I set the number of posts per page to 10, in the `config/initializers/kaminari_config.rb` file

chalmagean commented on app/views/admin/blog_posts/index.html.erb:
```diff
@@ -45,6 +45,8 @@
           <% end %>
         </table>
+        <% scope = BlogPost.all.page(params[:page]) %>
+        <%= paginate(scope) %>
```
You should use the instance variable here (`@blog_posts`) since that's what you want to paginate.

JakeWengroff replied:
Fixed. Also, as noted previously, no need to call `.all`.

chalmagean commented on app/views/blog_posts/index.html.erb:
```diff
@@ -14,6 +14,8 @@
       </div>
     <% end %>
+    <% scope = BlogPost.all.page(params[:page]) %>
+    <%= paginate(scope) %>
```
You should use the instance variable here (`@blog_posts`) since that's what you want to paginate.

JakeWengroff replied:
Done.

chalmagean commented on config/initializers/kaminari_config.rb:
```diff
@@ -0,0 +1,10 @@
+Kaminari.configure do |config|
+  config.default_per_page = 5
```
We can probably do 10 per page

JakeWengroff replied:
Done.

chalmagean commented on config/routes.rb:
```diff
@@ -6,7 +6,10 @@
     resources :blog_images
   end
+
+
```
Let's remove the extra space.

JakeWengroff replied:
Done. Again, bad habit.

chalmagean commented on spec/features/paginate_posts_index_spec.rb:
```diff
@@ -0,0 +1,14 @@
+  subject { page }
+  describe 'index is paginated' do
+    before do
+      25.times { FactoryGirl.create(:blog_post) }
```
Usually you want to avoid "magic numbers" like the 25 here. It doesn't tell you why that number is there or what does it represent. So what you want to do is give it a name like so:

```
POSTS_PER_PAGE = 25
POSTS_PER_PAGE.times { ... something ... }
```

Also it's preferable to use the same constant you use in the implementation code. For example you can define a constant in the `BlogPost` model like `BlogPost::POSTS_PER_PAGE` and you can use that everywhere (both implementation and spec).

chalmagean replied:
Also note that the `FactoryGirl` prefix is not required, you can just do `create(:blog_post)`.

JakeWengroff replied:
Done. These distinctions are important, and I'd like to learn more about defining constants.

chalmagean commented on spec/features/paginate_posts_index_spec.rb:
```diff
@@ -0,0 +1,14 @@
+    end
+      BlogPost.page(1).each do |post|
+        expect(page).to have_text('page')
+      end
```
You should have an `it` block that describes your test. For example:

```
it "limits the number of posts to #{BlogPost::POSTS_PER_PAGE}" do
  expect(page).to have_selector('.blog-post', count: BlogPost::POSTS_PER_PAGE)
end
```

chalmagean commented on spec/features/paginate_posts_index_spec.rb:
```diff
@@ -0,0 +1,14 @@
+      25.times { FactoryGirl.create(:blog_post) }
+      visit blog_posts_path
+    end
+      BlogPost.page(1).each do |post|
```
You don't really want to touch the inner workings of the app. You want to drive the user's browser in feature tests.

So you write feature tests by creating some context and then guiding the user  through the website and making assertions.

For example:

```
# Create more posts than what a page can hold
create_list(:blog_post, BlogPost::POSTS_PER_PAGE + 1)

# Drive the user
visit root_path
click_on "Blog"
expect(page).to have_selector('.blog-post', count: BlogPost::POSTS_PER_PAGE)
```

chalmagean commented on spec/features/paginate_posts_index_spec.rb:
```diff
@@ -0,0 +1,14 @@
+require 'spec_helper'
+
+describe 'Paginate posts index' do
+  subject { page }
```
I don't think you need this `subject` line

JakeWengroff replied:
You're right, it's gone now.

JakeWengroff committed:
Changes to the Feature test, Controller, Views.

I made suggested changes, but still need to revise the paginate posts spec file. The tests are all passing, however.

chalmagean commented on app/views/admin/blog_posts/index.html.erb:
```diff
@@ -45,6 +45,8 @@
             </tr>
           <% end %>
         </table>
+        <% @blog_posts = BlogPost.page(params[:page]) %>
```
`@blog_posts` is already set in the controller. You don't need to set it again in the view.

chalmagean commented on app/views/blog_posts/index.html.erb:
```diff
@@ -14,6 +14,8 @@
         <% end %>
       </div>
     <% end %>
+    <% @blog_posts = BlogPost.page(params[:page]) %>
```
`@blog_posts` is already set in the controller. You don't need to set it again in the view.

chalmagean commented on config/routes.rb:
```diff
@@ -6,6 +6,7 @@
   namespace :admin do
     resources :blog_posts do
       get "/get_tags" => "blog_posts#get_tags", as: :get_tags, on: :collection
+      get '/blog/:page', :action => :index, :on => :collection
```
Let's try to use the new hash syntax `key: :value` instead of `:key => :value`

JakeWengroff committed:
Addressed issues.

JakeWengroff commented:
Hi, Cezar,
I fixed the outstanding issues, but now `rspec features` is broken.  
Jake
